File: algo/combination-sum/combinationSum.py
class Solution(object):
    def combine(self, target, index, now, ret):
        for x in xrange(index, self.length):
            n = self.candidates[x]
            new_target = target - n
            if new_target == 0:
                ret.append(now + [n])
                return
            elif new_target < 0:
                return
            else:
                self.combine(new_target, x, now + [n], ret)
        
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.candidates = sorted(candidates)
        self.length = len(self.candidates)
        ret = []
        self.combine(target, 0, [], ret)
        return ret
        

File: algo/combination-sum/combinationSum.go
import "sort"

func combine(candidates *[]int, target int, index int, now []int, ret *([][]int)) {
    for i := index; i < len(*candidates); i++ {
        n := (*candidates)[i]
        new_target := target - n
        if new_target == 0 {
            now = append(now, n)
            *ret = append(*ret, now)
            break
        } else if new_target > 0 {
            now = append([]int{}, now...)
            combine(candidates, new_target, i, append(now, n), ret)
        } else {
            break
        }
    }
}

func combinationSum(candidates []int, target int) [][]int {
    ret := [][]int{}
    sort.Ints(candidates)
    combine(&candidates, target, 0, []int{}, &ret)
    return ret
}